## Data Collection

In [2]:
# Loading important libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
import sklearn
import warnings
warnings.filterwarnings('ignore')

In [6]:
# loading Dataset
train = pd.read_excel('Data_Train.xlsx')
test = pd.read_excel('Data_Test.xlsx')

## Data Exploration

In [8]:
train.shape, test.shape

((12690, 9), (4231, 8))

There are 12690 rows and 9 columns in train dataset and 4231 rows and 8 columns in test dataset.

In [9]:
train

TITLE  RESTAURANT_ID  \
0          CASUAL DINING           9438   
1      CASUAL DINING,BAR          13198   
2          CASUAL DINING          10915   
3            QUICK BITES           6346   
4         DESSERT PARLOR          15387   
...                  ...            ...   
12685        QUICK BITES          13228   
12686  CASUAL DINING,BAR           9686   
12687             LOUNGE          11133   
12688      CASUAL DINING           6134   
12689      CASUAL DINING           6430   

                                               CUISINES  \
0                           Malwani, Goan, North Indian   
1                        Asian, Modern Indian, Japanese   
2            North Indian, Chinese, Biryani, Hyderabadi   
3                                      Tibetan, Chinese   
4                                              Desserts   
...                                                 ...   
12685                       North Indian, Burger, Kebab   
12686                                 Goan, Continental   
12687          Finger Food, Continental, Asian, Chinese   
12688  North Indian, South Indian, Chinese, Street Food   
12689                             North Indian, Chinese   

                                                   TIME         CITY  \
0                11am – 4pm, 7:30pm – 11:30pm (Mon-Sun)        Thane   
1                                  6pm – 11pm (Mon-Sun)      Chennai   
2                   11am – 3:30pm, 7pm – 11pm (Mon-Sun)      Chennai   
3                               11:30am – 1am (Mon-Sun)       Mumbai   
4                                  11am – 1am (Mon-Sun)       Mumbai   
...                                                 ...          ...   
12685  12noon – 12midnight (Mon, Tue, Wed, Thu, Sun)...    Hyderabad   
12686   12noon – 1am (Mon-Fri),11am – 5pm, 7pm – 1am...       Mumbai   
12687                        12noon – 12:30AM (Mon-Sun)  Navi Mumbai   
12688                           6am – 10:45pm (Mon-Sun)      Chennai   
12689                          12:30pm – 11pm (Mon-Sun)    Hyderabad   

                   LOCALITY RATING       VOTES  COST  
0            Dombivali East    3.6    49 votes  1200  
1                 Ramapuram    4.2    30 votes  1500  
2                Saligramam    3.8   221 votes   800  
3               Bandra West    4.1    24 votes   800  
4               Lower Parel    3.8   165 votes   300  
...                     ...    ...         ...   ...  
12685            Gachibowli    3.8   546 votes   500  
12686  Bandra Kurla Complex    4.3  1214 votes  1800  
12687                 Vashi    4.0   608 votes  1300  
12688           Maduravoyal    3.5    32 votes   400  
12689       Vanasthalipuram    3.4   677 votes   600  

[12690 rows x 9 columns]

In [10]:
# Checking for missing values in train dataset
train.isnull().sum()

TITLE               0
RESTAURANT_ID       0
CUISINES            0
TIME                0
CITY              112
LOCALITY           98
RATING              2
VOTES            1204
COST                0
dtype: int64

There is Presence of null values in CITY,LOCALITY,RATING and VOTES.

In [15]:
# Checking for missing values in test dataset
test.isnull().sum()

TITLE              0
RESTAURANT_ID      0
CUISINES           0
TIME               0
CITY              35
LOCALITY          30
RATING             2
VOTES            402
dtype: int64

There is Presence of null values in CITY, LOCALITY,RATING and VOTES.

In [11]:
# Checking for Datatype
train.dtypes

TITLE            object
RESTAURANT_ID     int64
CUISINES         object
TIME             object
CITY             object
LOCALITY         object
RATING           object
VOTES            object
COST              int64
dtype: object

In [12]:
for i in train.columns:
    print("Unique values in", i, train[i].nunique())

Unique values in TITLE 113
Unique values in RESTAURANT_ID 11892
Unique values in CUISINES 4155
Unique values in TIME 2689
Unique values in CITY 359
Unique values in LOCALITY 1416
Unique values in RATING 32
Unique values in VOTES 1847
Unique values in COST 86


In [14]:
train.describe()

RESTAURANT_ID          COST
count   12690.000000  12690.000000
mean     7759.134121    655.252246
std      4504.874150    627.003540
min         0.000000     20.000000
25%      3863.250000    300.000000
50%      7740.000000    500.000000
75%     11688.750000    800.000000
max     15573.000000  14000.000000

## Data Pre-processing

In [16]:
# merging train and test dataset
df = train.append(test,ignore_index=True)

In [17]:
df

TITLE  RESTAURANT_ID  \
0          CASUAL DINING           9438   
1      CASUAL DINING,BAR          13198   
2          CASUAL DINING          10915   
3            QUICK BITES           6346   
4         DESSERT PARLOR          15387   
...                  ...            ...   
16916      CASUAL DINING           9057   
16917               None           1247   
16918        QUICK BITES           8617   
16919        QUICK BITES           6485   
16920        QUICK BITES           2421   

                                            CUISINES  \
0                        Malwani, Goan, North Indian   
1                     Asian, Modern Indian, Japanese   
2         North Indian, Chinese, Biryani, Hyderabadi   
3                                   Tibetan, Chinese   
4                                           Desserts   
...                                              ...   
16916                 North Indian, Mughlai, Chinese   
16917  Biryani, North Indian, Sandwich, Salad, Wraps   
16918                      Continental, North Indian   
16919                               Rolls, Beverages   
16920                                        Mughlai   

                                                    TIME         CITY  \
0                 11am – 4pm, 7:30pm – 11:30pm (Mon-Sun)        Thane   
1                                   6pm – 11pm (Mon-Sun)      Chennai   
2                    11am – 3:30pm, 7pm – 11pm (Mon-Sun)      Chennai   
3                                11:30am – 1am (Mon-Sun)       Mumbai   
4                                   11am – 1am (Mon-Sun)       Mumbai   
...                                                  ...          ...   
16916                        11:30am – 11:30pm (Mon-Sun)    New Delhi   
16917                               11am – 1am (Mon-Sun)    Bangalore   
16918                         9:30am – 10:30pm (Mon-Sun)    Faridabad   
16919  11am – 11:30pm (Mon, Tue, Wed, Thu, Sat, Sun),...        Kochi   
16920                                5pm – 1am (Mon-Sun)  Bandra West   

               LOCALITY RATING      VOTES    COST  
0        Dombivali East    3.6   49 votes  1200.0  
1             Ramapuram    4.2   30 votes  1500.0  
2            Saligramam    3.8  221 votes   800.0  
3           Bandra West    4.1   24 votes   800.0  
4           Lower Parel    3.8  165 votes   300.0  
...                 ...    ...        ...     ...  
16916      Punjabi Bagh    3.9  287 votes     NaN  
16917        HSR Layout    4.3  469 votes     NaN  
16918         Sector 86    3.7   53 votes     NaN  
16919             Kochi      -        NaN     NaN  
16920  Perry Cross Road    3.7   63 votes     NaN  

[16921 rows x 9 columns]

In [18]:
df = df[['TITLE', 'CUISINES', 'TIME', 'CITY', 'LOCALITY', 'RATING', 'VOTES', 'COST']]

In [19]:
import re
def extract_closed(time):
    a = re.findall('Closed \(.*?\)', time)
    if a != []:
        return a[0]
    else:
        return 'NA'

df['CLOSED'] = df['TIME'].apply(extract_closed)

In [20]:
df['TIME'] = df['TIME'].str.replace(r'Closed \(.*?\)','')
df['RATING'] = df['RATING'].str.replace('NEW', '1')
df['RATING'] = df['RATING'].str.replace('-', '1').astype(float)
df['VOTES'] = df['VOTES'].str.replace(' votes', '').astype(float)
df['CITY'].fillna('Missing', inplace=True)  
df['LOCALITY'].fillna('Missing', inplace=True)  
df['RATING'].fillna(3.8, inplace=True)  
df['VOTES'].fillna(0.0, inplace=True) 
df['COST'] = df['COST'].astype(float)

In [21]:
df.head(3)

TITLE                                    CUISINES  \
0      CASUAL DINING                 Malwani, Goan, North Indian   
1  CASUAL DINING,BAR              Asian, Modern Indian, Japanese   
2      CASUAL DINING  North Indian, Chinese, Biryani, Hyderabadi   

                                     TIME     CITY        LOCALITY  RATING  \
0  11am – 4pm, 7:30pm – 11:30pm (Mon-Sun)    Thane  Dombivali East     3.6   
1                    6pm – 11pm (Mon-Sun)  Chennai       Ramapuram     4.2   
2     11am – 3:30pm, 7pm – 11pm (Mon-Sun)  Chennai      Saligramam     3.8   

   VOTES    COST CLOSED  
0   49.0  1200.0     NA  
1   30.0  1500.0     NA  
2  221.0   800.0     NA

In [22]:
df['TITLE'].nunique(), df['CUISINES'].nunique()

(123, 5183)

In [23]:
calc_mean = df.groupby(['CITY'], axis=0).agg({'RATING': 'mean'}).reset_index()
calc_mean.columns = ['CITY','CITY_MEAN_RATING']
df = df.merge(calc_mean, on=['CITY'],how='left')

calc_mean = df.groupby(['LOCALITY'], axis=0).agg({'RATING': 'mean'}).reset_index()
calc_mean.columns = ['LOCALITY','LOCALITY_MEAN_RATING']
df = df.merge(calc_mean, on=['LOCALITY'],how='left')

In [24]:
df.head(2)

TITLE                        CUISINES  \
0      CASUAL DINING     Malwani, Goan, North Indian   
1  CASUAL DINING,BAR  Asian, Modern Indian, Japanese   

                                     TIME     CITY        LOCALITY  RATING  \
0  11am – 4pm, 7:30pm – 11:30pm (Mon-Sun)    Thane  Dombivali East     3.6   
1                    6pm – 11pm (Mon-Sun)  Chennai       Ramapuram     4.2   

   VOTES    COST CLOSED  CITY_MEAN_RATING  LOCALITY_MEAN_RATING  
0   49.0  1200.0     NA          3.376271              3.388889  
1   30.0  1500.0     NA          3.584588              3.472222

In [29]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf1 = TfidfVectorizer(ngram_range=(1, 1), lowercase=True)
df_title = tf1.fit_transform(df['TITLE'])
df_title = pd.DataFrame(data=df_title.toarray(), columns=tf1.get_feature_names_out())

tf2 = TfidfVectorizer(ngram_range=(1, 1), lowercase=True)
df_cuisines = tf2.fit_transform(df['CUISINES'])
df_cuisines = pd.DataFrame(data=df_cuisines.toarray(), columns=tf2.get_feature_names_out())

tf3 = TfidfVectorizer(ngram_range=(1, 1), lowercase=True)
df_city = tf3.fit_transform(df['CITY'])
df_city = pd.DataFrame(data=df_city.toarray(), columns=tf3.get_feature_names_out())

tf4 = TfidfVectorizer(ngram_range=(1, 1), lowercase=True)
df_locality = tf4.fit_transform(df['LOCALITY'])
df_locality = pd.DataFrame(data=df_locality.toarray(), columns=tf4.get_feature_names_out())

tf5 = TfidfVectorizer(ngram_range=(1, 1), lowercase=True)
df_time = tf5.fit_transform(df['TIME'])
df_time = pd.DataFrame(data=df_time.toarray(), columns=tf5.get_feature_names_out())

In [30]:
df.head(3)

TITLE                                    CUISINES  \
0      CASUAL DINING                 Malwani, Goan, North Indian   
1  CASUAL DINING,BAR              Asian, Modern Indian, Japanese   
2      CASUAL DINING  North Indian, Chinese, Biryani, Hyderabadi   

                                     TIME     CITY        LOCALITY  RATING  \
0  11am – 4pm, 7:30pm – 11:30pm (Mon-Sun)    Thane  Dombivali East     3.6   
1                    6pm – 11pm (Mon-Sun)  Chennai       Ramapuram     4.2   
2     11am – 3:30pm, 7pm – 11pm (Mon-Sun)  Chennai      Saligramam     3.8   

   VOTES    COST CLOSED  CITY_MEAN_RATING  LOCALITY_MEAN_RATING  
0   49.0  1200.0     NA          3.376271              3.388889  
1   30.0  1500.0     NA          3.584588              3.472222  
2  221.0   800.0     NA          3.584588              3.550000

In [31]:
df = pd.concat([df, df_title, df_cuisines, df_city, df_locality, df_time], axis=1) 
df.drop(['TITLE', 'CUISINES', 'CITY', 'LOCALITY', 'TIME'], axis=1, inplace=True)

In [32]:
df = pd.get_dummies(df, columns=['CLOSED'], drop_first=True)

In [33]:
df.shape

(16921, 2285)

In [34]:
train_df = df[df['COST'].isnull()!=True]
test_df = df[df['COST'].isnull()==True]
test_df.drop('COST', axis=1, inplace=True)

In [35]:
train_df.shape, test_df.shape

((12690, 2285), (4231, 2284))

In [36]:
train_df['COST'] = np.log1p(train_df['COST'])

## Train_Test_Split

In [37]:
X = train_df.drop(labels=['COST'], axis=1)
y = train_df['COST'].values

from sklearn.model_selection import train_test_split
X_train, X_cv, y_train, y_cv = train_test_split(X, y, test_size=0.25, random_state=1)

In [38]:
X_train.shape, y_train.shape, X_cv.shape, y_cv.shape

((9517, 2284), (9517,), (3173, 2284), (3173,))

## Building the Model

### Random Forest

In [47]:
from sklearn.ensemble import RandomForestRegressor

# Training Random Forest model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Predicting on the validation set
y_pred_rf = rf_model.predict(X_cv)

# Calculating RMSLE
rmsle_rf = np.sqrt(mean_squared_log_error(y_cv, y_pred_rf))
print('Random Forest RMSLE:', rmsle_rf)


Random Forest RMSLE: 0.05145556700339754


In [48]:
from sklearn.metrics import r2_score

r2 = r2_score(y_cv, y_pred_rf)

In [49]:
r2

0.7448143129134424